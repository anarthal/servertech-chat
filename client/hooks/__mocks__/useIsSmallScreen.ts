const fn = jest.fn();
fn.mockReturnValue(false);

export default fn;
