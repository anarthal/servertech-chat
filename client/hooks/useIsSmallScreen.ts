import { useState, useLayoutEffect } from "react";

const useIsSmallScreen = (): boolean => {
  const [matches, setMatches] = useState<boolean>(false);

  useLayoutEffect(() => {
    // Tailwind "md" breakpoint
    const media = matchMedia("(width < 48rem)");
    if (media.matches !== matches) {
      setMatches(media.matches);
    }

    // Listen to changes
    const listener = () => setMatches(media.matches);
    media.addEventListener("change", listener);

    // Remove the event listener when unmounting
    return () => media.removeEventListener("change", listener);
  }, [matches]);

  return matches;
};

export default useIsSmallScreen;
