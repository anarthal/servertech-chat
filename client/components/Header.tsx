import React from "react";
import Image from "next/image";
import boostLogo from "@/public/boost.png";
import ArrowBackIcon from "@mui/icons-material/ArrowBack";
import GitHubIcon from "@mui/icons-material/GitHub";
import useIsSmallScreen from "@/hooks/useIsSmallScreen";

// The common Header with the Boost logo shown in all pages

const links = [
  { text: "Source code", href: "https://github.com/anarthal/servertech-chat" },
  { text: "Docs", href: "https://anarthal.github.io/servertech-chat/" },
];

const BoostLogo = ({ height }: { height: number }) => {
  return (
    <a href="https://www.boost.org/">
      <Image src={boostLogo} height={height} alt="Boost logo"></Image>
    </a>
  );
};

// Header for large screens. Exported for testing
export const LargeHeader = () => {
  return (
    <div className="flex m-3">
      <BoostLogo height={60} />
      <div className="flex-1 flex justify-end align-middle">
        {links.map(({ text, href }) => (
          <div key={href} className="flex flex-col justify-center pr-12 pl-12">
            <a className="no-underline text-2xl text-black" href={href}>
              {text}
            </a>
          </div>
        ))}
      </div>
    </div>
  );
};

// Header for small screens. Exported for testing
export const SmallHeader = ({
  showArrow,
  onArrowClick,
}: {
  showArrow: boolean;
  onArrowClick?: () => void;
}) => {
  return (
    <div className="flex m-3">
      {showArrow && (
        <div
          onClick={onArrowClick}
          className="flex flex-col justify-center pr-2"
        >
          <ArrowBackIcon
            color={onArrowClick === undefined ? "disabled" : undefined}
          />
        </div>
      )}
      <div className="flex flex-1 justify-between">
        <BoostLogo height={40} />
        <a href="https://github.com/anarthal/servertech-chat" className="pl-2">
          <GitHubIcon style={{ width: "40px", height: "40px" }} />
        </a>
      </div>
    </div>
  );
};

export default function Header({
  showArrow,
  onArrowClick = undefined,
}: {
  showArrow: boolean;
  onArrowClick?: () => void;
}) {
  const isSmallScreen = useIsSmallScreen();
  return (
    <>
      {isSmallScreen ? (
        <SmallHeader showArrow={showArrow} onArrowClick={onArrowClick} />
      ) : (
        <LargeHeader />
      )}
    </>
  );
}
