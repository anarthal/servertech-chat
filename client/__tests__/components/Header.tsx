import { render, screen } from "@testing-library/react";
import Header, { SmallHeader, LargeHeader } from "@/components/Header";
import "@testing-library/jest-dom";
import useIsSmallScreen from "@/hooks/useIsSmallScreen";

jest.mock("@/hooks/useIsSmallScreen");

describe("Header", () => {
  test("Small header: arrow shown", () => {
    // Setup
    const fn = jest.fn();

    // Render
    const { asFragment } = render(
      <SmallHeader showArrow={true} onArrowClick={fn} />,
    );

    // The arrow is there
    const arrow = screen.getByTestId("ArrowBackIcon");
    expect(arrow).toBeInTheDocument();

    // Clicking the arrow triggers the function
    arrow.parentElement.click();
    expect(fn).toHaveBeenCalledTimes(1);

    // Snapshot test
    expect(asFragment()).toMatchSnapshot();
  });

  test("Small header: arrow shown disabled when callback is undefined", () => {
    // Render
    const { asFragment } = render(
      <SmallHeader showArrow={true} onArrowClick={undefined} />,
    );

    // The arrow is there
    const arrow = screen.getByTestId("ArrowBackIcon");
    expect(arrow).toBeInTheDocument();

    // Clicking the arrow does not crash
    arrow.parentElement.click();

    // Snapshot test
    expect(asFragment()).toMatchSnapshot();
  });

  test("Small header: arrow not shown", () => {
    // Render
    const { asFragment } = render(
      <SmallHeader showArrow={false} onArrowClick={undefined} />,
    );

    // The arrow is not there
    expect(screen.queryByTestId("ArrowBackIcon")).toBeNull();

    // Snapshot test
    expect(asFragment()).toMatchSnapshot();
  });

  test("Large header", () => {
    // Render
    const { asFragment } = render(<LargeHeader />);

    // Sanity check
    expect(screen.getByText("Source code")).toBeInTheDocument();

    // Snapshot test
    expect(asFragment()).toMatchSnapshot();
  });

  test("Header: large screen", () => {
    // Setup
    // @ts-ignore
    useIsSmallScreen.mockReturnValue(false);

    // Render
    render(<Header onArrowClick={undefined} showArrow={true} />);

    // The large header was rendered
    expect(screen.getByText("Source code")).toBeInTheDocument();
    expect(screen.queryByTestId("ArrowBackIcon")).toBeNull();
  });

  test("Header: small screen", () => {
    // Setup
    // @ts-ignore
    useIsSmallScreen.mockReturnValue(true);

    // Render
    render(<Header onArrowClick={undefined} showArrow={true} />);

    // The small header was rendered
    expect(screen.queryByText("Source code")).toBeNull();
    expect(screen.getByTestId("ArrowBackIcon")).toBeInTheDocument();
  });
});
