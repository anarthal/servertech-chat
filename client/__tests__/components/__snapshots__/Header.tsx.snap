// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Large header 1`] = `
<DocumentFragment>
  <div
    class="flex m-3"
  >
    <a
      href="https://www.boost.org/"
    >
      <img
        alt="Boost logo"
        data-nimg="1"
        decoding="async"
        height="60"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=128&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=64&q=75 1x, /_next/image?url=%2Fimg.jpg&w=128&q=75 2x"
        style="color: transparent;"
        width="60"
      />
    </a>
    <div
      class="flex-1 flex justify-end align-middle"
    >
      <div
        class="flex flex-col justify-center pr-12 pl-12"
      >
        <a
          class="no-underline text-2xl text-black"
          href="https://github.com/anarthal/servertech-chat"
        >
          Source code
        </a>
      </div>
      <div
        class="flex flex-col justify-center pr-12 pl-12"
      >
        <a
          class="no-underline text-2xl text-black"
          href="https://anarthal.github.io/servertech-chat/"
        >
          Docs
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Header Small header: arrow not shown 1`] = `
<DocumentFragment>
  <div
    class="flex m-3"
  >
    <div
      class="flex flex-1 justify-between"
    >
      <a
        href="https://www.boost.org/"
      >
        <img
          alt="Boost logo"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </a>
      <a
        class="pl-2"
        href="https://github.com/anarthal/servertech-chat"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="GitHubIcon"
          focusable="false"
          style="width: 40px; height: 40px;"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 1.27a11 11 0 00-3.48 21.46c.55.09.73-.28.73-.55v-1.84c-3.03.64-3.67-1.46-3.67-1.46-.55-1.29-1.28-1.65-1.28-1.65-.92-.65.1-.65.1-.65 1.1 0 1.73 1.1 1.73 1.1.92 1.65 2.57 1.2 3.21.92a2 2 0 01.64-1.47c-2.47-.27-5.04-1.19-5.04-5.5 0-1.1.46-2.1 1.2-2.84a3.76 3.76 0 010-2.93s.91-.28 3.11 1.1c1.8-.49 3.7-.49 5.5 0 2.1-1.38 3.02-1.1 3.02-1.1a3.76 3.76 0 010 2.93c.83.74 1.2 1.74 1.2 2.94 0 4.21-2.57 5.13-5.04 5.4.45.37.82.92.82 2.02v3.03c0 .27.1.64.73.55A11 11 0 0012 1.27"
          />
        </svg>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Header Small header: arrow shown 1`] = `
<DocumentFragment>
  <div
    class="flex m-3"
  >
    <div
      class="flex flex-col justify-center pr-2"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="ArrowBackIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20z"
        />
      </svg>
    </div>
    <div
      class="flex flex-1 justify-between"
    >
      <a
        href="https://www.boost.org/"
      >
        <img
          alt="Boost logo"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </a>
      <a
        class="pl-2"
        href="https://github.com/anarthal/servertech-chat"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="GitHubIcon"
          focusable="false"
          style="width: 40px; height: 40px;"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 1.27a11 11 0 00-3.48 21.46c.55.09.73-.28.73-.55v-1.84c-3.03.64-3.67-1.46-3.67-1.46-.55-1.29-1.28-1.65-1.28-1.65-.92-.65.1-.65.1-.65 1.1 0 1.73 1.1 1.73 1.1.92 1.65 2.57 1.2 3.21.92a2 2 0 01.64-1.47c-2.47-.27-5.04-1.19-5.04-5.5 0-1.1.46-2.1 1.2-2.84a3.76 3.76 0 010-2.93s.91-.28 3.11 1.1c1.8-.49 3.7-.49 5.5 0 2.1-1.38 3.02-1.1 3.02-1.1a3.76 3.76 0 010 2.93c.83.74 1.2 1.74 1.2 2.94 0 4.21-2.57 5.13-5.04 5.4.45.37.82.92.82 2.02v3.03c0 .27.1.64.73.55A11 11 0 0012 1.27"
          />
        </svg>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Header Small header: arrow shown disabled when callback is undefined 1`] = `
<DocumentFragment>
  <div
    class="flex m-3"
  >
    <div
      class="flex flex-col justify-center pr-2"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeMedium css-3nihpm-MuiSvgIcon-root"
        data-testid="ArrowBackIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20z"
        />
      </svg>
    </div>
    <div
      class="flex flex-1 justify-between"
    >
      <a
        href="https://www.boost.org/"
      >
        <img
          alt="Boost logo"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </a>
      <a
        class="pl-2"
        href="https://github.com/anarthal/servertech-chat"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="GitHubIcon"
          focusable="false"
          style="width: 40px; height: 40px;"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 1.27a11 11 0 00-3.48 21.46c.55.09.73-.28.73-.55v-1.84c-3.03.64-3.67-1.46-3.67-1.46-.55-1.29-1.28-1.65-1.28-1.65-.92-.65.1-.65.1-.65 1.1 0 1.73 1.1 1.73 1.1.92 1.65 2.57 1.2 3.21.92a2 2 0 01.64-1.47c-2.47-.27-5.04-1.19-5.04-5.5 0-1.1.46-2.1 1.2-2.84a3.76 3.76 0 010-2.93s.91-.28 3.11 1.1c1.8-.49 3.7-.49 5.5 0 2.1-1.38 3.02-1.1 3.02-1.1a3.76 3.76 0 010 2.93c.83.74 1.2 1.74 1.2 2.94 0 4.21-2.57 5.13-5.04 5.4.45.37.82.92.82 2.02v3.03c0 .27.1.64.73.55A11 11 0 0012 1.27"
          />
        </svg>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
